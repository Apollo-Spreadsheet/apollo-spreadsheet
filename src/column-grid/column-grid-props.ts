import {StretchMode} from "../types/stretch-mode.enum"
import {HeadersData} from "./types/header.type";
import {GridTheme} from "../types/grid-theme";
import {NavigationCoords} from "../navigation/types/navigation-coords.type";

export interface ColumnGridProps {
	headers: HeadersData;
	className?: string;
	minRowHeight: number;

	defaultColumnWidth: number
	getColumnWidth: ({index}: { index: number}) => number;

	width: number;
	scrollLeft: number;
	isScrolling: boolean;
	onScroll: Function;
	height: number;
	theme?: GridTheme;
	coords: NavigationCoords;
	/** @default StretchMode.None  */
	stretchMode?: StretchMode
}