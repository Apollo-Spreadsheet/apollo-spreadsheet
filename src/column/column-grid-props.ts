import {StretchMode} from "./stretch-mode.enum"
import {HeadersData} from "./types/header.type";
import {GridTheme} from "../types/grid-theme";
import {NavigationCoords} from "../navigation/types/navigation-coords.type";

export interface ColumnGridProps {
	/** @default StretchMode.None  */
	stretchMode?: StretchMode
	headers: HeadersData;
	className: string;
	minRowHeight: number;
	columnWidth: number;
	autoHeight: boolean;
	width: number;
	scrollLeft: number;
	isScrolling: boolean;
	onScroll: Function;
	height: number;
	theme?: GridTheme;
	coords: NavigationCoords;
}