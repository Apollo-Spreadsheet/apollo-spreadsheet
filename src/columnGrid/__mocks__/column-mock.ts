import { Header } from '../types/header.type'

export const createColumnMock = (column?: Partial<Header>) => {
	return {
		id: 'column - ' + Math.random(),
		accessor: column?.accessor ?? 'access-' + Math.random(),
		...column,
	} as Header
}
