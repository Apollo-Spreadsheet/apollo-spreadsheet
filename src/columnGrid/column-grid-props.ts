import { StretchMode } from '../types/stretch-mode.enum'
import { GridHeader, Column, NestedHeader } from './types/header.type'
import { GridTheme } from '../types/grid-theme'
import { NavigationCoords } from '../navigation/types/navigation-coords.type'
import { DisableSortFilterParam } from '../gridWrapper/gridWrapperProps'
import GridWrapper from '../gridWrapper/GridWrapper'
import React from 'react'
import { ApiRef } from '../api/types'

export interface ColumnGridProps {
	/**
	 * Indicates if the sort is disabled globally or on a specific column
	 * @default true **/
	disableSort?: boolean | DisableSortFilterParam
	apiRef: ApiRef
	data: Array<GridHeader[]>
	columns: Column[]
	nestedHeaders?: Array<NestedHeader[]>
	minRowHeight: number

	defaultColumnWidth: number
	getColumnWidth: ({ index }: { index: number }) => number

	width: number
	scrollLeft: number
	/** @default StretchMode.None  */
	stretchMode?: StretchMode
	/**
	 * Overscan count buffer for react-virtualized
	 * @description Keep in mind a lower value
	 * @default 2
	 */
	overscanRowCount?: number
	/**
	 * Overscan count buffer for react-virtualized
	 * @description Keep in mind a lower value
	 * @default 2
	 */
	overscanColumnCount?: number
}
