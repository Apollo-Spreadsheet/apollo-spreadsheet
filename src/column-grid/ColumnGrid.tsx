import React, {
	useRef,
	useEffect,
	useMemo,
	useCallback,
	useImperativeHandle,
	forwardRef,
} from "react";
import {Grid, CellMeasurerCache} from "react-virtualized";
import CellMeasurer from "../core/CellMeasureWrapper";
import useLazyRef from "../hooks/useLazyRef";
import {Column} from "./types/header.type";
import {insertDummyCells} from "../utils/helpers";
import clsx from "clsx";
import {ColumnGridProps} from "./column-grid-props";

const ColumnGrid = React.memo(
  forwardRef((props: ColumnGridProps, componentRef) => {
	  const {current: cache} = useLazyRef(
		() =>
		  new CellMeasurerCache({
			  defaultWidth: props.defaultColumnWidth,
			  defaultHeight: props.minRowHeight,
			  fixedWidth: true,
			  minHeight: props.minRowHeight,
			  //We might use another approach in here
			  minWidth: props.defaultColumnWidth,
		  })
	  );

	  useImperativeHandle(componentRef, () => ({
		  recomputeGridSize: () => {
			  gridRef.current?.recomputeGridSize();
		  },
		  forceUpdate: () => {
			  gridRef.current?.forceUpdate();
		  },
	  }));
	  const gridRef = useRef<Grid | null>(null);

	  const data = useMemo(() => {
		  return insertDummyCells(props.headers);
	  }, [props.headers]);

	  // clear cache and recompute when data changes
	  useEffect(() => {
		  cache?.clearAll();
		  gridRef.current?.recomputeGridSize();
	  }, [data]);


	  const headerRendererWrapper = useCallback(
		({style, cell, ref, columnIndex, rowIndex}) => {
			const {title, renderer} = cell as Column;
			/** @todo Cache cell renderer result because if may have not changed so no need to invoke again **/
			const children = renderer ? (renderer(cell) as any) : title;

			return (
			  <div
				ref={ref}
				/** @todo We ensure its the first rowIndex to avoid nested getting this but i have to discuss with Pedro **/
				className={
					props.coords.colIndex === columnIndex && rowIndex === 0
					  ? clsx(
					  props.theme?.headerClass,
					  props.theme?.currentColumnClass
					  )
					  : props.theme?.headerClass
				}
				style={{
					display: "flex",
					justifyContent: "center",
					padding: "5px",
					boxSizing: "border-box",
					background: "#efefef",
					border: "1px solid #ccc",
					cursor: "default",
					...style,
				}}
			  >
				  {children}
			  </div>
			);
		},
		[props.coords, props.theme, props.width]
	  );

	  const cellMeasurerWrapperRenderer = useCallback(
		(args) => {
			const cell = data[args.rowIndex]?.[args.columnIndex];
			if (!cell) {
				return null;
			}
			const style = {
				...args.style,
				...cell["style"],
				width: props.getColumnWidth({index: args.columnIndex}),
				userSelect: "none",
			};

			const rendererProps = {
				...args,
				cell,
			};
			return (
			  <CellMeasurer
				cache={cache}
				columnIndex={args.columnIndex}
				key={args.key}
				parent={args.parent}
				rowIndex={args.rowIndex}
				rowSpan={cell.rowSpan}
				colSpan={cell.colSpan}
				cellRenderer={headerRendererWrapper}
				rendererProps={rendererProps}
				style={style}
			  />
			);
		},
		[data, props.theme, props.coords, props.width]
	  );

	  const columnCount = useMemo(() => {
		  return data.length ? data[0].length : 0;
	  }, [data]);

	  const onRefMount = useCallback((instance) => {
		  gridRef.current = instance;
	  }, []);

	  return (
		<Grid
		  {...(props as any)}
		  ref={onRefMount}
		  cellRenderer={cellMeasurerWrapperRenderer}
		  deferredMeasurementCache={cache}
		  rowHeight={cache.rowHeight}
		  rowCount={data.length}
		  columnCount={columnCount}
		  overscanRowCount={2}
		  overscanColumnCount={2}
		  width={props.width}
		  columnWidth={props.getColumnWidth}
		  autoHeight
		/>
	  );
  })
);

export default ColumnGrid;
