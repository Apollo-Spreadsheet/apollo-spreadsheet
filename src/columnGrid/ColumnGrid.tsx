import React, {
	useRef,
	useEffect,
	useMemo,
	useCallback,
	useImperativeHandle,
	forwardRef,
} from 'react'
import { Grid, CellMeasurerCache } from 'react-virtualized'
import CellMeasurer from '../cellMeasurer/CellMeasureWrapper'
import { Column } from './types/header.type'
import clsx from 'clsx'
import { ColumnGridProps } from './column-grid-props'
import { insertDummyCells } from '../core/utils/insertDummyCells'
import { MeasurerRendererProps } from '../cellMeasurer/cellMeasureWrapperProps'
import Tooltip from '@material-ui/core/Tooltip'

export const ColumnGrid = React.memo(
	forwardRef((props: ColumnGridProps, componentRef) => {
		const cache = useRef(
			new CellMeasurerCache({
				defaultWidth: props.defaultColumnWidth,
				defaultHeight: props.minRowHeight,
				fixedWidth: true,
				minHeight: props.minRowHeight,
				//We might use another approach in here
				minWidth: props.defaultColumnWidth,
			}),
		).current

		useImperativeHandle(componentRef, () => ({
			recomputeGridSize: () => {
				gridRef.current?.recomputeGridSize()
			},
			forceUpdate: () => {
				gridRef.current?.forceUpdate()
			},
		}))
		const gridRef = useRef<Grid | null>(null)

		/** @todo Review insertDummyCells parsing relative to flatMap but also the single array parse is not working well **/
		const data = useMemo(() => {
			return insertDummyCells(props.headers)
		}, [props.headers])

		// clear cache and recompute when data changes
		useEffect(() => {
			cache?.clearAll()
			gridRef.current?.recomputeGridSize()
		}, [data])

		const headerRendererWrapper = useCallback(
			({ style, cell, ref, columnIndex, rowIndex }) => {
				const { title, renderer } = cell as Column
				/** @todo Cache cell renderer result because if may have not changed so no need to invoke again **/
				const children = renderer ? (
					(renderer(cell) as any)
				) : cell.tooltip ? (
					<Tooltip title={title} arrow placement={'top'} {...cell.tooltipProps}>
						<span>{title}</span>
					</Tooltip>
				) : (
					title
				)

				//Ensure dummy cells doesn't have any styling
				const headerClassName =
					!cell.dummy && props.coords.colIndex === columnIndex && rowIndex === 0
						? clsx(props.theme?.headerClass, props.theme?.currentColumnClass)
						: !cell.dummy
						? props.theme?.headerClass
						: undefined

				/**
				 * @todo If it is a nested header we need to load the styling from the theme property and combine using clsx
				 */
				return (
					<div
						ref={ref}
						className={headerClassName}
						style={{
							display: 'flex',
							justifyContent: 'center',
							padding: '5px',
							boxSizing: 'border-box',
							background: '#efefef',
							border: '1px solid #ccc',
							cursor: 'default',
							...style,
						}}
					>
						{children}
					</div>
				)
			},
			[props.coords, props.theme, props.width],
		)

		const cellMeasurerWrapperRenderer = useCallback(
			args => {
				const cell = data[args.rowIndex]?.[args.columnIndex]
				if (!cell) {
					return null
				}
				const style = {
					...args.style,
					//TODO Review this style property
					...cell['style'],
					width: props.getColumnWidth({ index: args.columnIndex }),
					userSelect: 'none',
				}

				const rendererProps: MeasurerRendererProps = {
					...args,
					cell,
					getColumnWidth: props.getColumnWidth,
				}
				return (
					<CellMeasurer
						cache={cache}
						columnIndex={args.columnIndex}
						key={args.key}
						parent={args.parent}
						rowIndex={args.rowIndex}
						rowSpan={cell.rowSpan}
						colSpan={cell.colSpan}
						cellRenderer={headerRendererWrapper}
						rendererProps={rendererProps}
						style={style}
					/>
				)
			},
			[data, props.theme, props.coords, props.width],
		)

		const columnCount = useMemo(() => {
			return data.length ? data[0].length : 0
		}, [data])

		const onRefMount = useCallback(instance => {
			gridRef.current = instance
		}, [])

		return (
			<Grid
				{...props}
				ref={onRefMount}
				cellRenderer={cellMeasurerWrapperRenderer}
				deferredMeasurementCache={cache}
				rowHeight={cache.rowHeight}
				rowCount={data.length}
				columnCount={columnCount}
				overscanRowCount={2}
				overscanColumnCount={2}
				width={props.width}
				columnWidth={props.getColumnWidth}
				autoHeight
			/>
		)
	}),
)

export default ColumnGrid
