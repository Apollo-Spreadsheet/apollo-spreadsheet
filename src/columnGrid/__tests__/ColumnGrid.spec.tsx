import { shallow } from 'enzyme'
import ColumnGrid from '../ColumnGrid'
import { ColumnGridProps } from '../column-grid-props'
import React from 'react'
import { createColumnMock } from '../__mocks__/column-mock'

describe('<ColumnGrid />', () => {
	const props: ColumnGridProps = {
		headers: [[
			createColumnMock({ id: '1', width: 100, title: 'First' }),
			createColumnMock({ id: '12', title: 'Second', tooltip: 'Test' }),
		]],
		minRowHeight: 10,
		defaultColumnWidth: 10,
		getColumnWidth: ({ index }: { index: number }) => 0,
		width: 100,
		scrollLeft: 0,
		isScrolling: false,
		height: 100,
		coords: { rowIndex: 0, colIndex: 0 },
	}
	const grid = shallow(<ColumnGrid {...props} />)

	it('should match the snapshot', () => {
		expect(grid.html()).toMatchSnapshot()
	})

})