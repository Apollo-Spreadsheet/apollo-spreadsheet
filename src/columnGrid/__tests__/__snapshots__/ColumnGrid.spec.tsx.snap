// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColumnGrid /> should match the snapshot 1`] = `"<div aria-label=\\"grid\\" aria-readonly=\\"true\\" class=\\"ReactVirtualized__Grid\\" role=\\"grid\\" style=\\"box-sizing:border-box;direction:ltr;height:auto;position:relative;width:100px;-webkit-overflow-scrolling:touch;will-change:transform;overflow-x:hidden;overflow-y:hidden\\" tabindex=\\"0\\"><div class=\\"ReactVirtualized__Grid__innerScrollContainer\\" role=\\"rowgroup\\" style=\\"width:0;height:10px;max-width:0;max-height:10px;overflow:hidden;pointer-events:;position:relative\\"><div class=\\"\\" style=\\"display:flex;justify-content:center;padding:5px;box-sizing:border-box;background:#efefef;border:1px solid #ccc;cursor:default;height:auto;left:0;position:absolute;top:0;width:0;user-select:none;transform:translate3d(0, 0, 0);align-items:center;overflow:hidden;word-break:break-word;text-overflow:ellipsis;text-align:center\\">First</div><div style=\\"display:flex;justify-content:center;padding:5px;box-sizing:border-box;background:#efefef;border:1px solid #ccc;cursor:default;height:auto;left:0;position:absolute;top:0;width:0;user-select:none;transform:translate3d(0, 0, 0);align-items:center;overflow:hidden;word-break:break-word;text-overflow:ellipsis;text-align:center\\"><span title=\\"Second\\" class=\\"\\">Second</span></div></div></div>"`;
